src/main/java/org/alliancegenome/curation_api/services/validation/ReagentValidator.java

```diff
@@ -0,0 +1,99 @@
+
+public class ReagentValidator extends AuditedObjectValidator<Reagent> {
+
+	@Inject
+	NcbiTaxonTermService ncbiTaxonTermService;
+	@Inject
+	DataProviderService dataProviderService;
+	@Inject
+	DataProviderValidator dataProviderValidator;
+	@Inject
+	DataProviderDAO dataProviderDAO;
+
+	public Reagent validateCommonReagentFields(Reagent uiEntity, Reagent dbEntity) {
+		
+		Boolean newEntity = false;
+		if (dbEntity.getId() == null)
+			newEntity = true;
+		dbEntity = validateAuditedObjectFields(uiEntity, dbEntity, newEntity);
+		
+		String modEntityId = StringUtils.isNotBlank(uiEntity.getModEntityId()) ? uiEntity.getModEntityId() : null;
+		dbEntity.setModEntityId(modEntityId);
+		
+		String modInternalId = StringUtils.isNotBlank(uiEntity.getModInternalId()) ? uiEntity.getModInternalId() : null;
+		dbEntity.setModInternalId(modInternalId);
+	
+		DataProvider dataProvider = validateDataProvider(uiEntity, dbEntity);
+		dbEntity.setDataProvider(dataProvider);
+		
+		NCBITaxonTerm taxon = validateTaxon(uiEntity, dbEntity);
+		dbEntity.setTaxon(taxon);
+		
+		return dbEntity;
+	}
+	
+	public NCBITaxonTerm validateTaxon(Reagent uiEntity, Reagent dbEntity) {
+		String field = "taxon";
+		if (uiEntity.getTaxon() == null || StringUtils.isBlank(uiEntity.getTaxon().getCurie()))
+			return null;
+
+		NCBITaxonTerm taxon = ncbiTaxonTermService.get(uiEntity.getTaxon().getCurie()).getEntity();
+		if (taxon == null) {
+			addMessageResponse(field, ValidationConstants.INVALID_MESSAGE);
+			return null;
+		}
+		
+		if (taxon.getObsolete() && (dbEntity.getTaxon() == null || !taxon.getCurie().equals(dbEntity.getTaxon().getCurie()))) {
+			addMessageResponse(field, ValidationConstants.OBSOLETE_MESSAGE);
+			return null;
+		}
+		
+		return taxon;
+	}
+	
+	public DataProvider validateDataProvider(Reagent uiEntity, Reagent dbEntity) {
+		String field = "dataProvider";
+		
+		if (uiEntity.getDataProvider() == null) {
+			if (dbEntity.getId() == null)
+				uiEntity.setDataProvider(dataProviderService.createAffiliatedModDataProvider());
+			if (uiEntity.getDataProvider() == null) {
+				addMessageResponse(field, ValidationConstants.REQUIRED_MESSAGE);
+				return null;
+			}
+		} 
+
+		DataProvider uiDataProvider = uiEntity.getDataProvider();
+		DataProvider dbDataProvider = dbEntity.getDataProvider();
+		
+		ObjectResponse<DataProvider> dpResponse = dataProviderValidator.validateDataProvider(uiDataProvider, dbDataProvider, false);
+		if (dpResponse.hasErrors()) {
+			addMessageResponse(field, dpResponse.errorMessagesString());
+			return null;
+		}
+		
+		DataProvider validatedDataProvider = dpResponse.getEntity();
+		if (validatedDataProvider.getObsolete() && (dbDataProvider == null || !validatedDataProvider.getId().equals(dbDataProvider.getId()))) {
+			addMessageResponse(field, ValidationConstants.OBSOLETE_MESSAGE);
+			return null;
+		}
+		
+		if (validatedDataProvider.getId() == null)
+			validatedDataProvider = dataProviderDAO.persist(validatedDataProvider);
```

Comment:
Ideally you would use the dataProvider service here.

Reply:
Refactored